packages/drivers/web-utils/src/utils.ts

```diff
@@ -1,13 +1,77 @@
 import getElementCategory, { tags } from "./getElementCategory";
 import isElementHidden from "./isElementHidden";
-import { ElementCategory } from "./types";
+import {
+  ElementCategory,
+  SelectorCriteria,
+  InterestingAttributes,
+} from "./types";
+import { CRITERIA_CONFIG } from "./config";
 
 declare global {
   interface Window {
     createMarkedViewHierarchy: () => string;
+    findElement: (criteria: SelectorCriteria) => HTMLElement | null;
   }
 }
 
+/**
+ * Retrieves the “interesting” attributes from an element.
+ * For `rect`, it uses the element’s top-left (x, y) coordinates.
+ */
+export function getInterestingAttributes(
+  element: HTMLElement,
+): InterestingAttributes {
+  const attributes = {} as InterestingAttributes;
+  for (const key in CRITERIA_CONFIG) {
+    const criteriaKey = key as keyof InterestingAttributes;
+    const config = CRITERIA_CONFIG[criteriaKey];
+    attributes[criteriaKey] = config.extract(element);
+  }
+  return attributes;
+}
+
+export function findElement(criteria: SelectorCriteria): HTMLElement | null {
```

Comment:
you can adopt more functional approach here, high-level overview of how I see it

1. Short-circuits (for `critical` matchers), currently only aria-label
2. Extract error-cal logic to a new method (`calcCandidateError`) to compute the total error for each candidate. Within that method, you can use `reduce()` to sum the error on the criteria keys and return Infinity if an enforced attribute fails.
3. Use `map()` on candidates to turn them into `{candidate, errorScore}` pairs by calling `calcCandidateError`.
4. Use `reduce()` on those pairs to pick the one with the lowest error.
5. Return the best candidate if it's lower than max error score threshold, or null if none is found.